### PR TITLE
HIVE-19219: Hive replicated database is out of sync if events are cleaned-up.

### PR DESCRIPTION
**Pull request by sankarh**

### CODE REVIEW

File: standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java [hunk 1]
```diff
@@ -2587,19 +2587,24 @@ public NotificationEventResponse getNextNotification(long lastEventId, int maxEv
     rqst.setMaxEvents(maxEvents);
     NotificationEventResponse rsp = client.get_next_notification(rqst);
     LOG.debug("Got back " + rsp.getEventsSize() + " events");
```
Comment (maheshk114) [inline standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java]: if response can be null ..this log can cause exception

File: standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java [hunk 2]
```diff
@@ -2587,19 +2587,24 @@ public NotificationEventResponse getNextNotification(long lastEventId, int maxEv
...
-      NotificationEventResponse filtered = new NotificationEventResponse();
-      if (rsp != null && rsp.getEvents() != null) {
-        for (NotificationEvent e : rsp.getEvents()) {
-          if (filter.accept(e)) {
-            filtered.addToEvents(e);
-          }
+    NotificationEventResponse filtered = new NotificationEventResponse();
+    if (rsp != null && rsp.getEvents() != null) {
+      long nextEventId = lastEventId + 1;
+      for (NotificationEvent e : rsp.getEvents()) {
```
Comment (maheshk114) [inline standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java]: for this to work properly, the select should be done ordered by event id

File: standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java [hunk 3]
```diff
@@ -2587,19 +2587,24 @@ public NotificationEventResponse getNextNotification(long lastEventId, int maxEv
...
+        if (e.getEventId() != nextEventId) {
```
[the remark this answers is not in the document]
Comment (sankarh) [reply standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java]: Planning to change the logic to check if the first event is there even after we dump all the events. So, this contiguous event id check would be removed. Do you think any problem with this logic?

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java [hunk 4]
```diff
@@ -808,73 +861,27 @@ public NotificationEventResponse apply(@Nullable NotificationEventResponse event
...
-                    event.setEventId(nextEventId++);
-                  }
-                  break;
-                }
-              }
-
-              outEventIds.add(event);
+            // Skip all the INSERT events
+            if (event.getDbName().equalsIgnoreCase(dbName) && event.getEventType().equalsIgnoreCase("INSERT")) {
+              injectionPathCalled = true;
```
[the remark this answers is not in the document]
Comment (sankarh) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java]: It is used to check if my query hit this method or not. It is validated somewhere below after REPL DUMP.

File: standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java (hunk 1 above)
Comment (sankarh) [reply standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java]: rsp won't be null.

File: standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java (hunk 2 above)
Comment (sankarh) [reply standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java]: Yes, metastore call will fetch it in ascending order.

**Pull request closed without merging**